api/src/main/java/io/minio/MinioClient.java

```diff
@@ -1509,6 +1535,107 @@ public void getObject(String bucketName, String objectName, String fileName)
   }
 
   /**
+   * Gets entire encrypted object's data as {@link InputStream} in given bucket, then returns CipherInputStream.
+   * CipherInputStream is composed of an InputStream and a Cipher so that read() methods return data that are read in
+   * from the underlying InputStream but have been additionally processed by the Cipher. The Cipher is initialized for
+   * decryption, the CipherInputStream will attempt to read in data and decrypt them, before returning the decrypted
+   * data.
+   * <p>
+   * The CipherInputStream must be closed after use else the connection will remain open.
+   * </p>
+   * <b>Example:</b>
+   *
+   * <pre>
+   * {
+   *   // Generate encryption materials
+   *   KeyGenerator symKeyGenerator = KeyGenerator.getInstance("AES");
+   *   symKeyGenerator.init(256);
+   *   SecretKey symKey = symKeyGenerator.generateKey();
+   *   EncryptionMaterials encMaterials = new EncryptionMaterials(symKey);
+   *
+   *   // Get object
+   *   InputStream stream = minioClient.getObject("my-bucketname", "my-objectname", encMaterials);
+   *   byte[] buf = new byte[16384];
+   *   int bytesRead;
+   *   while ((bytesRead = stream.read(buf, 0, buf.length)) >= 0) {
+   *     System.out.println(new String(buf, 0, bytesRead));
+   *   }
+   *   stream.close();
+   * }
+   * </pre>
+   *
+   * @param bucketName
+   *          Bucket name.
+   * @param objectName
+   *          Object name in the bucket.
+   * @param encryptionMaterials
+   *          Encryption metadata details and operations.
+   *
+   * @return {@link InputStream} containing the object data.
+   *
+   * @throws InvalidBucketNameException
+   *           upon invalid bucket name is given
+   * @throws NoResponseException
+   *           upon no response from server
+   * @throws IOException
+   *           upon connection error
+   * @throws XmlPullParserException
+   *           upon parsing response xml
+   * @throws ErrorResponseException
+   *           upon unsuccessful execution
+   * @throws InternalException
+   *           upon internal library error
+   * @throws InvalidEncryptionMetadataException
+   *           upon encryption key/iv error
+   * @throws BadPaddingException
+   *           upon wrong padding
+   * @throws IllegalBlockSizeException
+   *           upon incorrect block size
+   * @throws NoSuchPaddingException
+   *           upon incorrect padding
+   * @throws InvalidAlgorithmParameterException
+   *           upon incorrect algorithm
+   */
+  public CipherInputStream getObject(String bucketName, String objectName, EncryptionMaterials encryptionMaterials)
```

Comment:
Its like 
```java
public void getObject(String bucketName, String objectName, String fileName);
public InputStream getObject(String bucketName, String objectName);
public InputStream getObject(String bucketName, String objectName, long offset);
public InputStream getObject(String bucketName, String objectName, long offset, Long length);
```

These methods uses primarily one method which is `public InputStream getObject(String bucketName, String objectName, long offset, Long length);`, rest of other methods are just wrappers.

WRT encrypt/decrypt methods, we would need below
```java
public InputStream getObject(String bucketName, String objectName, SecretKey key);
public InputStream getObject(String bucketName, String objectName, SecretKey key, Cipher cipher);
public InputStream getObject(String bucketName, String objectName, KeyPair key);
public InputStream getObject(String bucketName, String objectName, KeyPair key, Cipher cipher);
```

I think its hard to implement below
```java
public InputStream getObject(String bucketName, String objectName, long offset, Long length,
                             SecretKey key, Cipher cipher);
public InputStream getObject(String bucketName, String objectName, long offset, Long length,
                             KeyPair key, Cipher cipher);
```

Similarly for `putObject()`
```java
public void putObject(String bucketName, String objectName, String fileName, String contentType, SecretKey key,
                      Cipher cipher);
public void putObject(String bucketName, String objectName, String fileName, String contentType, KeyPair key,
                      Cipher cipher);
public void putObject(String bucketName, String objectName, InputStream stream, long size, String contentType,
                      Map<String, String> headerMap, SecretKey key, Cipher cipher);
public void putObject(String bucketName, String objectName, InputStream stream, long size, String contentType,
                      Map<String, String> headerMap, KeyPair key, Cipher cipher);
```